### PR TITLE
Test split artwork shard

### PR DESCRIPTION
**Pull request by aslakhellesoy**
This adds a Mocha unit test for `splitArtworkShard`.

### CODE REVIEW

**Comment by eivindingebrigtsen** [comment]
@kristaki90 @StefPler 

We have pushed some changes to this branch with a test that sometimes:
SCR-20230619-w3j
works and most times fails 
SCR-20230619-w3t


Is there some event we need to wait for, when issuing the ArtworkShard, 
before we can split or transfer the shard to another user? 

We added a delay of 5 seconds, which I saw you did in the publish script, that's when it suddenly started working once in a while 🦺

File: .envrc [hunk 1]
```diff
@@ -6,3 +6,11 @@ export PUBLISHER_ID="0x7bd8a9757894069470bdee4d899b34b41ed82cbd77f68ba33d0fd46d0
 export USER1_PHRASE="gossip net bamboo assist danger armed soda plastic such father better grid"
 export USER2_PHRASE="can gaze pistol praise unable man afraid include route doll evoke ticket"
 export USER3_PHRASE="inherit chat inmate swamp transfer deposit shy heart frost chest cage bonus"
+export ADMIN_CAP_ID="0xa7785dd1cf60421de8e1a9dea7910e7e5615f25e26bebf7610c77aa9663a604f"
```
Comment (aslakhellesoy) [inline .envrc:9]: @eivindingebrigtsen looks like the build just keeps appending to this file. Should we:

1. Add it to `.gitignore`
2. Make sure it doesn't append but overwrite

File: test/splitArtworkShard.test.ts [hunk 2]
```diff
@@ -10,9 +12,10 @@ describe("splitArtworkShard", () => {
     artworkId = await mintArtwork(mintArtworkOptions);
   });
 
-  it("should split an artwork shard", async () => {
+  it.only("should split an artwork shard", async () => {
     const artworkShardId = await mintArtworkShard({ artworkId, phrase: USER1_PHRASE, shares: 10 });
 
-    await splitArtworkShard(artworkShardId, 2);
+    const splitArtworkShardId = await splitArtworkShard(artworkShardId, 2);
+    assert.ok(splitArtworkShardId);
```
[the remark this answers is not in the document]
Comment (aslakhellesoy) [reply test/splitArtworkShard.test.ts:21]: I think a more thorough assertion would be to check that we actually have two shards now, and check that they both belong to the same user.

**Comment by StefPler** [comment]
> 

@eivindingebrigtsen This is an issue that can happen when interacting with a load balanced RPC urls. The issue might be that the block has not propagated yet to the Fullnode you are trying to query when creating the new transaction to split the artwork shard.
What is the `parent_id` in this scenario? (if this is Testnet or Devnet, providing text of the objects involved in the error would be great)

Something you could try is: Use the TS sdk `waitForTransactionBlock` method and provide the create shard tx digest as input before proceeding to make sure the transaction has propagated. A sleep could also work but its not recommended

**Comment by kristaki90** [comment]
> > waitForTransactionBlock
> 
> @eivindingebrigtsen you need to add `requestType: "WaitForLocalExecution",` you can check this [here](https://docs.sui.io/sui-jsonrpc#sui_executeTransactionBlock)

As I checked this is already there though @StefPler

**Comment by StefPler** [comment]
> thank you @StefPler - I cannot seem to find any documentation on a `waitForTransactionBlock` method on sui.io
> Can you point me in a direction to an example ? 

Unfortunately we don't have a documention for that function yet. However I can share with you the following example:
```ts
  public async hasBlockSynced(digest: string) { 
     if (!digest) throw new Error("Invalid txn digest"); 
  
     try { 
       let resp = await this.provider.waitForTransactionBlock({ 
         timeout: 20000, 
         digest, 
         options: { showEffects: true }, 
       }); 
  
       return resp?.effects?.status.status === "success"; 
     } catch (e) { 
       console.error("sync error", e); 
       return false; 
     } 
   }
```

You can get the digest from any transaction response

**Comment by aslakhellesoy** [comment]
Thanks @StefPler that's super helpful!

**Comment by aslakhellesoy** [comment]
Thanks @StefPler - we're still have a puzzle here.

> You can get the digest from any transaction response.

There are many digests in the response from the `mint_artwork_shard` function. Here is an example:

```json
{
  "digest": "F9HFT8Yw4zxYMd9qwPZzft99hZwT6e84XBwnJcYHj4ww",
  "effects": {
    "messageVersion": "v1",
    "status": {
      "status": "success"
    },
    "executedEpoch": "0",
    "gasUsed": {
      "computationCost": "1000000",
      "storageCost": "11552000",
      "storageRebate": "5808528",
      "nonRefundableStorageFee": "58672"
    },
    "modifiedAtVersions": [
      {
        "objectId": "0x52ec344599422718a5683c45f2b3019b77735977985ef627324d5d3c76154dfb",
        "sequenceNumber": "42"
      },
      {
        "objectId": "0x77b14d98b08759738790ef5f3470add269823385c625d7fcc71f1a06f0dd5eb3",
        "sequenceNumber": "42"
      },
      {
        "objectId": "0xbb33ace3e303650b315fa515468c575338da562a0379e3901291f68124747afb",
        "sequenceNumber": "42"
      }
    ],
    "sharedObjects": [
      {
        "objectId": "0xbb33ace3e303650b315fa515468c575338da562a0379e3901291f68124747afb",
        "version": 42,
        "digest": "HZCguhcwuVniPRmcPUWSbTfZheSqBinSQqXVRZPYkvY9"
      }
    ],
    "transactionDigest": "F9HFT8Yw4zxYMd9qwPZzft99hZwT6e84XBwnJcYHj4ww",
    "created": [
      {
        "owner": {
          "AddressOwner": "0x104df955a36761cc8b5d34b514c741aa225223bb615d1e23494d93a79c0fa06d"
        },
        "reference": {
          "objectId": "0x0e7968d88ce4498b5b4e21e9eb6dfe9c8d9bd47f1a2a8d76f0fb2443a37834f9",
          "version": 43,
          "digest": "DPMVbYAG3tRbaJp34j2WPa26aqZ4Z56uMedz4pzS1EXu"
        }
      },
      {
        "owner": {
          "ObjectOwner": "0xbb33ace3e303650b315fa515468c575338da562a0379e3901291f68124747afb"
        },
        "reference": {
          "objectId": "0x120d9fbf252f9098eb97befe8d38b4411398554515bbe25013a219a574f66983",
          "version": 43,
          "digest": "FTvZoxmaZRxe96MxZiWHTHmaykhCdyJBseVfh6Drc5HR"
        }
      }
    ],
    "mutated": [
      {
        "owner": {
          "AddressOwner": "0x9659dc2bb6036cba28e3f162cdecb39bd431a30fe13d9546d2ac1d8e20be9972"
        },
        "reference": {
          "objectId": "0x52ec344599422718a5683c45f2b3019b77735977985ef627324d5d3c76154dfb",
          "version": 43,
          "digest": "2tzAbAmpx3ziirG2nTiDX5guumgkJ2xwo9gX6dFec47n"
        }
      },
      {
        "owner": {
          "AddressOwner": "0x9659dc2bb6036cba28e3f162cdecb39bd431a30fe13d9546d2ac1d8e20be9972"
        },
        "reference": {
          "objectId": "0x77b14d98b08759738790ef5f3470add269823385c625d7fcc71f1a06f0dd5eb3",
          "version": 43,
          "digest": "HubYT7BfgoZPHHBx9YPm8mUE5CAUSxzrXFdZvszhKB4F"
        }
      },
      {
        "owner": {
          "Shared": {
            "initial_shared_version": 42
          }
        },
        "reference": {
          "objectId": "0xbb33ace3e303650b315fa515468c575338da562a0379e3901291f68124747afb",
          "version": 43,
          "digest": "7HMFD1UkdUmUiNHf2hHrVeK2NoGJzXStBnQegSBHdame"
        }
      }
    ],
    "gasObject": {
      "owner": {
        "AddressOwner": "0x9659dc2bb6036cba28e3f162cdecb39bd431a30fe13d9546d2ac1d8e20be9972"
      },
      "reference": {
        "objectId": "0x52ec344599422718a5683c45f2b3019b77735977985ef627324d5d3c76154dfb",
        "version": 43,
        "digest": "2tzAbAmpx3ziirG2nTiDX5guumgkJ2xwo9gX6dFec47n"
      }
    },
    "dependencies": [
      "9yhZwALnxL6axrzuxBkJVG1pGvSpNhvGLW9oNYUatq46",
      "GcSZe69iSrGLr5DSBS9TLqTBDpEaQj35C8mt9tB5Xrcu"
    ]
  },
  "objectChanges": [
    {
      "type": "mutated",
      "sender": "0x9659dc2bb6036cba28e3f162cdecb39bd431a30fe13d9546d2ac1d8e20be9972",
      "owner": {
        "AddressOwner": "0x9659dc2bb6036cba28e3f162cdecb39bd431a30fe13d9546d2ac1d8e20be9972"
      },
      "objectType": "0x2::coin::Coin<0x2::sui::SUI>",
      "objectId": "0x52ec344599422718a5683c45f2b3019b77735977985ef627324d5d3c76154dfb",
      "version": "43",
      "previousVersion": "42",
      "digest": "2tzAbAmpx3ziirG2nTiDX5guumgkJ2xwo9gX6dFec47n"
    },
    {
      "type": "mutated",
      "sender": "0x9659dc2bb6036cba28e3f162cdecb39bd431a30fe13d9546d2ac1d8e20be9972",
      "owner": {
        "AddressOwner": "0x9659dc2bb6036cba28e3f162cdecb39bd431a30fe13d9546d2ac1d8e20be9972"
      },
      "objectType": "0x1bdaf85b118942e927d356138032d589d1f68950f16dd178259b61a40437e92a::open_art_market::AdminCap",
      "objectId": "0x77b14d98b08759738790ef5f3470add269823385c625d7fcc71f1a06f0dd5eb3",
      "version": "43",
      "previousVersion": "42",
      "digest": "HubYT7BfgoZPHHBx9YPm8mUE5CAUSxzrXFdZvszhKB4F"
    },
    {
      "type": "mutated",
      "sender": "0x9659dc2bb6036cba28e3f162cdecb39bd431a30fe13d9546d2ac1d8e20be9972",
      "owner": {
        "Shared": {
          "initial_shared_version": 42
        }
      },
      "objectType": "0x1bdaf85b118942e927d356138032d589d1f68950f16dd178259b61a40437e92a::open_art_market::Artwork",
      "objectId": "0xbb33ace3e303650b315fa515468c575338da562a0379e3901291f68124747afb",
      "version": "43",
      "previousVersion": "42",
      "digest": "7HMFD1UkdUmUiNHf2hHrVeK2NoGJzXStBnQegSBHdame"
    },
    {
      "type": "created",
      "sender": "0x9659dc2bb6036cba28e3f162cdecb39bd431a30fe13d9546d2ac1d8e20be9972",
      "owner": {
        "AddressOwner": "0x104df955a36761cc8b5d34b514c741aa225223bb615d1e23494d93a79c0fa06d"
      },
      "objectType": "0x1bdaf85b118942e927d356138032d589d1f68950f16dd178259b61a40437e92a::open_art_market::ArtworkShard",
      "objectId": "0x0e7968d88ce4498b5b4e21e9eb6dfe9c8d9bd47f1a2a8d76f0fb2443a37834f9",
      "version": "43",
      "digest": "DPMVbYAG3tRbaJp34j2WPa26aqZ4Z56uMedz4pzS1EXu"
    },
    {
      "type": "created",
      "sender": "0x9659dc2bb6036cba28e3f162cdecb39bd431a30fe13d9546d2ac1d8e20be9972",
      "owner": {
        "ObjectOwner": "0xbb33ace3e303650b315fa515468c575338da562a0379e3901291f68124747afb"
      },
      "objectType": "0x2::dynamic_field::Field<address, 0x1bdaf85b118942e927d356138032d589d1f68950f16dd178259b61a40437e92a::open_art_market::Shares>",
      "objectId": "0x120d9fbf252f9098eb97befe8d38b4411398554515bbe25013a219a574f66983",
      "version": "43",
      "digest": "FTvZoxmaZRxe96MxZiWHTHmaykhCdyJBseVfh6Drc5HR"
    }
  ],
  "confirmedLocalExecution": true
}
```

We have tried `provider.waitForTransactionBlock` with the top level `digest`, but we're still seeing intermittent failures in the test. Are we using the wrong digest?

**Comment by eivindingebrigtsen** [comment]
Thank you @StefPler for figuring out the issue here. 
I’m currently on holiday with a mobile phone as my device, so will get back to the code in a weeks time.

I know it’s none of your faults, but shouldn’t there be a simpler way of deriving which object IDs was created with a transaction? 
`getCreatedObjects(txRes)?.find((obj: any) => obj?.owner?.AddressOwner === address)?.reference.objectId;` feels like a very complicated way of finding the ID of the newly created object. I’m sure that others are trying to find these IDs as well 😁

**Comment by StefPler** [comment]
> Thank you @StefPler for figuring out the issue here. I’m currently on holiday with a mobile phone as my device, so will get back to the code in a weeks time.
> 
> I know it’s none of your faults, but shouldn’t there be a simpler way of deriving which object IDs was created with a transaction? `getCreatedObjects(txRes)?.find((obj: any) => obj?.owner?.AddressOwner === address)?.reference.objectId;` feels like a very complicated way of finding the ID of the newly created object. I’m sure that others are trying to find these IDs as well grin

Hi there @eivindingebrigtsen , yes this feels a bit indirect. It would make more sense if we could chose an object created based on its type as well. Is this what you were thinking of as well? Maybe a lib method similar to `getCreatedObjects` that you can use, for example: `getCreatedObjectWithFilter(...args)`

**Pull request merged**